src/NodaTime.Test/DateIntervalTest.cs

```diff
@@ -169,6 +171,423 @@ public void Equals_DifferentToOtherType()
+
+            IEqualityComparer<DateInterval> comparer = DateInterval.ContainedDatesEqualityComparer;
+            Assert.IsFalse(comparer.Equals(interval1, interval2));
+            Assert.AreNotEqual(comparer.GetHashCode(interval1), comparer.GetHashCode(interval2));  // probably
+        }
+
+        [Test]
+        public void ContainedDatesEqualityComparer_Extremes_Unequal_DifferentCalendars()
+        {
+            LocalDate start1 = new LocalDate(2000, 1, 1);
+            LocalDate end1 = new LocalDate(2001, 6, 19);
+            LocalDate start2 = start1.WithCalendar(CalendarSystem.Gregorian);
+            LocalDate end2 = end1.WithCalendar(CalendarSystem.Gregorian);
+            var interval1 = new DateInterval(start1, end1, true);
+            var interval2 = new DateInterval(start2, end2, true);
+            Assert.AreNotEqual(interval1, interval2);
+
+            // Neither interval above can be represented as an exclusive interval, but they must still compare unequal.
+            IEqualityComparer<DateInterval> comparer = DateInterval.ContainedDatesEqualityComparer;
+            Assert.IsFalse(comparer.Equals(interval1, interval2));
+            Assert.AreNotEqual(comparer.GetHashCode(interval1), comparer.GetHashCode(interval2));  // probably
+        }
+
+        [Test]
+        [TestCaseSource(nameof(SupportedCalendars))]
+        public void ContainedDatesEqualityComparer_EmptyRange(CalendarSystem calendar)
+        {
+            int year = (calendar.MaxYear + calendar.MinYear) / 2;
+            LocalDate date1 = new LocalDate(year, 2, 1, calendar);
+            LocalDate date2 = new LocalDate(year, 2, 2, calendar);
+            var interval1 = new DateInterval(date1, date1, false);
+            var interval2 = new DateInterval(date2, date2, false);
+            Assert.AreEqual(0, interval2.Length);
+            Assert.AreEqual(0, interval1.Length);
+            Assert.AreNotEqual(interval1, interval2);
+
+            // This is the key difference between this comparer and the normalizing comparer.
+            IEqualityComparer<DateInterval> comparer = DateInterval.ContainedDatesEqualityComparer;
+            Assert.IsTrue(comparer.Equals(interval1, interval2));
+            Assert.AreEqual(comparer.GetHashCode(interval1), comparer.GetHashCode(interval2));
+        }
+
+        [Test]
+        public void ContainedDatesEqualityComparer_EmptyRange_DifferentCalendars()
+        {
+            LocalDate date1 = new LocalDate(2000, 2, 1);
+            LocalDate date2 = new LocalDate(2000, 2, 2, CalendarSystem.Gregorian);
+            var interval1 = new DateInterval(date1, date1, false);
+            var interval2 = new DateInterval(date2, date2, false);
+            Assert.AreEqual(0, interval2.Length);
+            Assert.AreEqual(0, interval1.Length);
+            Assert.AreNotEqual(interval1, interval2);
+
+            // Different calendar systems are still unequal, even if they're both empty.
```

Comment:
At some point we could *consider* having yet another equality comparer (or two) which effectively normalized calendars as well. (Every non-ISO date can be converted to an ISO date, so it would be simple to normalize both intervals to ISO ones, and then compare.)